tools/codegen/cmd/featuregate-test-analyzer.go

```diff
@@ -0,0 +1,284 @@
+		if !ok {
+			errs = append(errs, fmt.Errorf("%q in %q on %v: missing testing data for ", featureGate, clusterProfile, requiredVariant))
+			missingVariants[requiredVariant] = true
+		}
+	}
+
+	for jobVariant, testedVariant := range testingResults {
+		if _, ok := missingVariants[jobVariant]; ok {
+			continue
+		}
+		if len(testedVariant.TestResults) < requiredNumberOfTests {
+			errs = append(errs, fmt.Errorf("%q in %q on %v: only %d tests found, need at least %d", featureGate, clusterProfile, jobVariant, len(testedVariant.TestResults), requiredNumberOfTests))
+		}
+		for _, testResults := range testedVariant.TestResults {
+			if testResults.TotalRuns < requiredNumberOfTestRunsPerVariantInTwoWeeks {
+				errs = append(errs, fmt.Errorf("%q in %q on %v: %q only has %d runs, need at least %d runs", featureGate, clusterProfile, jobVariant, testResults.TestName, testResults.TotalRuns, requiredNumberOfTestRunsPerVariantInTwoWeeks))
+			}
+			if testResults.TotalRuns == 0 {
+				continue
+			}
+			passPercent := float32(testResults.SuccessfulRuns) / float32(testResults.TotalRuns)
+			if passPercent < requiredPassRateOfTestsPerVariant {
+				displayExpected := int(requiredPassRateOfTestsPerVariant * 100)
+				displayActual := int(passPercent * 100)
+				errs = append(errs, fmt.Errorf("%q in %q on %v: %q only passed %d%%, need at least %d%%", featureGate, clusterProfile, jobVariant, testResults.TestName, displayActual, displayExpected))
+			}
+		}
+	}
+
+	return errs
+}
+
+var (
+	requiredSelfManagedJobVariants = []JobVariant{
+		{
+			Cloud:        "aws",
+			Architecture: "amd64",
+			Topology:     "HA",
+		},
+		{
+			Cloud:        "aws",
+			Architecture: "amd64",
+			Topology:     "single-node",
+		},
+		{
+			Cloud:        "azure",
+			Architecture: "amd64",
+			Topology:     "HA",
+		},
+		{
+			Cloud:        "gcp",
+			Architecture: "amd64",
+			Topology:     "HA",
+		},
+		{
+			Cloud:        "metal",
+			Architecture: "amd64",
+			Topology:     "HA",
+		},
+		{
+			Cloud:        "metal",
+			Architecture: "amd64",
+			Topology:     "single-node",
+		},
+		{
+			Cloud:        "vsphere",
+			Architecture: "amd64",
+			Topology:     "HA",
+		},
+	}
+
+	requiredHypershiftJobVariants = []JobVariant{
+		{
+			Cloud:        "aws",
+			Architecture: "amd64",
+			Topology:     "hypershift",
+		},
+		// ibm and powervs?
+	}
+)
+
+type JobVariant struct {
+	Cloud        string
+	Architecture string
+	Topology     string
+}
+
+type TestingResults struct {
+	JobVariant JobVariant
+
+	TestResults []TestResults
+}
+
+type TestResults struct {
+	TestName       string
+	TotalRuns      int
+	SuccessfulRuns int
+	FailedRuns     int
+	FlakedRuns     int
+}
+
+func listTestResultFor(clusterProfile, featureGate string) (map[JobVariant]*TestingResults, error) {
+	testPattern := fmt.Sprintf("[OCPFeatureGate:%s]", featureGate)
+	fmt.Printf("Query component readiness for all test run results for pattern %q on clusterProfile %q\n", testPattern, clusterProfile)
```

Comment:
@dgoodwin @stbenjam this is the spot where I'd like to query component readiness and/or sippy for data and then process the returned data to produce something like the map being returned from this method.

Reply:
We had a discussion during the team sync about the card related to this -- I think capabilities are a better fit for component readiness than using test pattern matching.  It's useful beyond this effort, for example if I wanted to see a `Disruption` dashboard, disruption isn't present in all the test names, but they *do* all have the right capability.

Capabilities can span components, and a test can have more than one capability.

https://github.com/openshift-eng/ci-test-mapping/pull/114 adds the capability ("Feature Gate: X") to tests marked with `[OCPFeatureGate:]`

Reply:
While we work on the alternative component readiness API's you asked for, you can get the data you want already from Sippy.  

For example, here's aws amd64 for existing `Feature:NetworkPolicy` tests:

http://sippy.dptools.openshift.org/api/tests?release=4.16&filter=%7B%22items%22%3A%5B%7B%22columnField%22%3A%22variants%22%2C%22not%22%3Afalse%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22aws%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22not%22%3Afalse%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22amd64%22%7D%2C%7B%22columnField%22%3A%22name%22%2C%22not%22%3Afalse%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22NetworkPolicy%22%7D%5D%2C%22linkOperator%22%3A%22and%22%7D&period=default&sortField=current_working_percentage&sort=asc&collapse=true

The filter model used in the URL is just this urlencoded:

```json
{
  "items": [
    {
      "columnField": "variants",
      "not": false,
      "operatorValue": "contains",
      "value": "aws"
    },
    {
      "columnField": "variants",
      "not": false,
      "operatorValue": "contains",
      "value": "amd64"
    },
    {
      "columnField": "name",
      "not": false,
      "operatorValue": "contains",
      "value": "NetworkPolicy"
    }
  ],
  "linkOperator": "and"
}
```

Reply:
@stbenjam Please be sure that capability mechanism you're using will accommodate both upstream and downstream tests.  (notice two patterns [here](https://github.com/openshift/origin/pull/28694/files#diff-e57b70f3677d343b31b83c28349a03e9dcce3ff5bb14e1977e45a2cd371deef9R152-R155)).  Not all upstream featuregates are controlled using OCP featuresets.  Though I could probably live over-subdivision.

Reply:
It looks like Stephen captured both FeatureGate and OCPFeatureGate in https://github.com/openshift-eng/ci-test-mapping/pull/114

The test mapping repo change added FeatureGate as a capability in the mapping table:
```
{
  "kind": "TestOwnership",
  "apiVersion": "v1",
  "id": "openshift-tests:ec05afb1e465ba7a8813abbf42ae36ab",
  "name": "[sig-api-machinery] ValidatingAdmissionPolicy [Privileged:ClusterAdmin] [FeatureGate:ValidatingAdmissionPolicy] [Beta] should allow expressions to refer variables. [Suite:openshift/conformance/parallel] [Suite:k8s]",
  "suite": "openshift-tests",
  "product": "OpenShift",
  "component": "kube-apiserver",
  "capabilities": ["FeatureGate:ValidatingAdmissionPolicy"],
  "staff_approved_obsolete": "false",
  "priority": "-1",
  "created_at": "2024-04-10T04:00:15.391627",
  "jira_component": "kube-apiserver",
  "jira_component_id": "12367637"
}
```

After https://github.com/openshift/sippy/pull/1603 lands, you can query with[ this API](http://localhost:8080/api/component_readiness?baseEndTime=2024-02-28T23:59:59Z&baseRelease=4.15&baseStartTime=2024-02-01T00:00:00Z&confidence=95&excludeArches=arm64,heterogeneous,ppc64le,s390x&excludeClouds=openstack,ibmcloud,libvirt,ovirt,unknown&excludeVariants=hypershift,osd,microshift,techpreview,single-node,assisted,compact&groupBy=cloud,arch,network&ignoreDisruption=true&ignoreMissing=false&minFail=3&pity=5&sampleEndTime=2024-04-08T23:59:59Z&sampleRelease=4.16&sampleStartTime=2024-04-02T00:00:00Z&capability=FeatureGate:ValidatingAdmissionPolicy), and get the following result:

```
{
    "rows": [
        {
            "component": "kube-apiserver",
            "columns": [
                {
                    "network": "ovn",
                    "arch": "amd64",
                    "platform": "aws",
                    "status": 1
                },
                {
                    "network": "ovn",
                    "arch": "amd64",
                    "platform": "vsphere",
                    "status": 1
                }
            ]
        }
    ],
    "generated_at": "2024-04-10T11:30:56.587081-04:00"
}
```

You can the components/variants where there is test coverage. Because of the missing basis, the status is set to 1. This is where a comparison with a fixed threshold might help.